File: java_spring-boot_spring-web-mvc_spring-data-jpa_lombok/src/main/java/org/example/demoapp/employee/EmployeeService.java
package org.example.demoapp.service;

import javax.transaction.Transactional;

import java.util.List;
import java.util.Optional;

import org.example.demoapp.entity.Employee;
import org.example.demoapp.repository.EmployeeRepository;
import org.springframework.stereotype.Service;

@Service
@Transactional
public class EmployeeService {

    private final EmployeeRepository employeeRepository;

    public EmployeeService(EmployeeRepository employeeRepository) {
        this.employeeRepository = employeeRepository;
    }

    public List<Employee> findAll() {
        return employeeRepository.findAll();
    }

    public Optional<Employee> findById(long id) {
        return employeeRepository.findById(id);
    }

    public Employee save(Employee employee) {
        return employeeRepository.save(employee);
    }

    public void deleteById(long id) {
        employeeRepository.findById(id)
                .ifPresent(employeeRepository::delete);
    }

    public void updateEmployee(long id, Employee employee) {
        Employee existingEmployee = employeeRepository.findById(id)
                .orElseThrow(EmployeeNotFoundException::new);
        existingEmployee.updateFrom(employee);
    }
}

File: kotlin_spring-boot_spring-boot-data-jpa_spring-web_maven/src/main/kotlin/org/example/demoapp/employee/EmployeeService.kt
package org.example.demoapp.employee

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
class EmployeeService(val employeeRepository: EmployeeRepository) {

    fun findAll(): List<Employee> {
        return employeeRepository.findAll()
    }

    fun findById(id: Long): Optional<Employee> {
        return employeeRepository.findById(id)
    }

    fun save(employee: Employee): Employee {
        return employeeRepository.save(employee)
    }

    fun updateEmployee(id: Long, updatedEmployee: Employee) {
        val employee = employeeRepository.findById(id)
                .orElseThrow { EmployeeNotFoundException() }
        employee.updateFrom(updatedEmployee)
    }

    fun delete(id: Long) {
        employeeRepository.findById(id)
                .ifPresent { employee: Employee -> employeeRepository.delete(employee) }
    }

}